pallets/inactivity-tracking/src/lib.rs

```diff
@@ -0,0 +1,258 @@
+        type SessionEndChecker: GetRandomnessForNextBlock<BlockNumberFor<Self>>;
+    }
+
+    /// A list of double map of inactive collators for a session
+    #[pallet::storage]
+    pub type ActiveCollators<T: Config> = StorageMap<
+        _,
+        Twox64Concat,
+        SessionIndex,
+        BoundedVec<T::CollatorId, T::MaxCollatorsPerSession>,
+        ValueQuery,
+    >;
+
+    /// A list of inactive collators for a session. Repopulated at the start of every session
+    #[pallet::storage]
+    pub type ActiveCollatorsForCurrentSession<T: Config> =
+        StorageValue<_, BoundedVec<T::CollatorId, T::MaxCollatorsPerSession>, ValueQuery>;
+
+    /// A list of inactive container chains for a session. Repopulated at the start of every session
+    #[pallet::storage]
+    pub type ActiveContainerChainsForCurrentSession<T: Config> =
+        StorageValue<_, BoundedVec<ParaId, T::MaxInactiveSessions>, ValueQuery>;
+
+    /// The last session index for which the inactive collators have not been processed
+    #[pallet::storage]
+    pub type LastUnprocessedSession<T: Config> = StorageValue<_, SessionIndex, ValueQuery>;
+
+    #[pallet::event]
+    pub enum Event<T: Config> {}
+
+    #[pallet::error]
+    pub enum Error<T> {
+        MaxCollatorsPerSessionReached,
+        MaxContainerChainsReached,
+    }
+
+    #[pallet::call]
+    impl<T: Config> Pallet<T> {}
+
+    #[pallet::hooks]
+    impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+        fn on_initialize(_n: BlockNumberFor<T>) -> Weight {
+            let mut total_weight = T::DbWeight::get().reads_writes(1, 0);
+
+            // Process the orchestrator chain block author (if it exists)
+            if let Some(orchestrator_chain_author) = T::GetSelfChainBlockAuthor::get_block_author()
+            {
+                total_weight += Self::on_author_noted(orchestrator_chain_author);
+            }
+
+            // Update inactive collator records only after a session has ended
+            let current_session = T::CurrentSessionIndex::session_index();
+            let current_unprocessed_session = <LastUnprocessedSession<T>>::get();
+            if current_unprocessed_session < current_session {
+                total_weight +=
+                    Self::process_ended_session(current_unprocessed_session, current_session);
+            }
+            total_weight
+        }
+
+        fn on_finalize(n: BlockNumberFor<T>) {
+            if T::SessionEndChecker::should_end_session(n) {
+                Self::process_inactive_chains_for_session();
+            }
+        }
+    }
+
+    impl<T: Config> Pallet<T> {
+        fn process_ended_session(
+            unprocessed_session_id: SessionIndex,
+            current_session_id: SessionIndex,
+        ) -> Weight {
+            let mut total_weight = T::DbWeight::get().reads_writes(0, 4);
+            ActiveCollators::<T>::insert(
+                unprocessed_session_id,
```

Comment:
this is wrong since you are here passing unprocessed_session_id, but the active collators have already changed. I am assuming that what you want to do is `current_session_id`

Reply:
No worries! Will be resolving this conversation and  moving all related changes to https://github.com/moondance-labs/tanssi/tree/aleks-collator-inactivity-tracking